### PR TITLE
Post api for a submission

### PR DESCRIPTION
**Pull request by shell-x**

### CODE REVIEW

**Comment by shell-x** [comment]
API is working as expected. With all checks passing similar to submission behaviour over UI. 

image

image

File: app/controllers/submissions_controller.rb [hunk 1]
```diff
@@ -93,7 +95,7 @@ def new
   end
 
   def create
-    session_info = {participant_id: current_participant.id, online_submission: true}
+    session_info = {participant_id: current_participant.id, online_submission: request.path.include?("api/v1") ? false : true}
```
Comment (shell-x) [inline app/controllers/submissions_controller.rb:98]: @skbly7 I made this field to false, if submission is from API. Any thoughts?

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -142,4 +142,14 @@ def default_url_options
       { host: ENV['DOMAIN_NAME'] || 'localhost:3000' }
     end
   end
+
+  def set_user_by_token(mapping = nil)
+    if request.path.split('/')[1]=='api'
```
Comment (shell-x) [inline app/controllers/application_controller.rb:147]: @skbly7 any better behaviour of checking if request is from API or UI?

File: app/controllers/submissions_controller.rb (hunk 1 above)
Comment (shell-x) [reply app/controllers/submissions_controller.rb:98]: submistted_received_from? web or api?

File: app/controllers/submissions_controller.rb [hunk 3]
```diff
@@ -225,12 +241,26 @@ def check_participation_terms
...
+      else
+        redirect_to [challenge, ParticipationTerms.current_terms]
+      end
       return
     end
 
     unless policy(challenge).has_accepted_challenge_rules?
-      redirect_to challenge_challenge_rules_path(challenge)
+      if request.path.include?("api/v1")
+        link_terms = "#{request.env["HTTP_HOST"]}#{challenge_challenge_rules_path(challenge)}"
```
Comment (skbly7) [inline app/controllers/submissions_controller.rb:257]: Use this for link instead: `challenge_challenge_rules_path(challenge)`

File: app/controllers/submissions_controller.rb [hunk 4]
```diff
@@ -283,29 +313,44 @@ def submission_params
...
   def handle_code_based_submissions
-    return if params[:submission][:submission_type] != "code"
-    file_location = get_s3_key.gsub("${filename}", "editor_input." + params[:language])
-    begin
-      S3_BUCKET.object(file_location).put(body: params[:submission][:submission_data])
-    rescue Aws::S3::Errors::ServiceError
-      redirect_to helpers.challenge_submissions_path(@challenge), notice: 'Submission upload failed, please try again.'
-    end
 
+    if request.path.include?("api/v1")
```
Comment (skbly7) [inline app/controllers/submissions_controller.rb:328]: Can we skip duplicated code here?

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (shell-x) [reply app/controllers/application_controller.rb:147]: streamline this check with modularity

File: app/controllers/submissions_controller.rb [hunk 5]
```diff
@@ -283,29 +313,44 @@ def submission_params
...
+      end
+    end
     params[:submission].merge!({
       submission_files_attributes: {"0": {seq: "", submission_type: params[:language], submission_file_s3_key: file_location}}
     })
   end
 
   def handle_artifact_based_submissions
-    return if params.dig('submission', 'submission_files_attributes', '0', 'submission_type') != 'artifact'
+    return if is_api_request ||
```
Comment (skbly7) [inline app/controllers/submissions_controller.rb:352]: ?

File: app/controllers/submissions_controller.rb [hunk 6]
```diff
@@ -225,12 +241,26 @@ def check_participation_terms
...
+        payload = { message: "Please accept Participant terms first here => #{link_terms}" }
+        status  = :ok
+        render json: payload, status: status
+      else
+        redirect_to [challenge, ParticipationTerms.current_terms]
+      end
       return
     end
 
     unless policy(challenge).has_accepted_challenge_rules?
```
Comment (skbly7) [inline app/controllers/submissions_controller.rb:255]: ```unless policy(challenge).has_accepted_challenge_rules? || policy(challenge).has_accepted_participation_terms?```

File: app/controllers/submissions_controller.rb [hunk 7]
```diff
@@ -225,12 +241,26 @@ def check_participation_terms
     end
 
     unless policy(challenge).has_accepted_participation_terms?
```
Comment (skbly7) [inline app/controllers/submissions_controller.rb:243]: You can remove this `unless`

File: app/controllers/submissions_controller.rb (hunk 5 above)
Comment (shell-x) [reply app/controllers/submissions_controller.rb:352]: done

**Comment by skbly7** [comment]
```
POST /api/v1/submissions

{
  challenge_id: "learning-to-smell",
  description: "I love AIcrowd <3",
  submission_files: [{
    "submission_file_s3_key": "https://aicrowd-production.s3.aws.com/..../yolo/abc.zip"
  },
  { ... }
  ]
}
```

```
GET /api/v1/submissions/new

{
  challenge_id: "learning-to-smell"
}

Response:
{
  success: true,
  data: {
     presigned_key: "https://aicrowd-production.s3.aws.com/..../yolo/abc.zip?token=23heu923hr2u3h"
  },
  message: "Presigned key generated successfully"
}
```

File: app/controllers/submissions_controller.rb [hunk 8]
```diff
@@ -159,11 +162,26 @@ def export
...
   end
 
+  def new_api
+    check_participation_terms
+    @submissions_remaining, @reset_dttm = SubmissionsRemainingQuery.new(
+      challenge:      @challenge,
+      participant_id: current_participant.id
+    ).call
+    if @submissions_remaining < 1
+      redirect_or_json(helpers.challenge_submissions_path(@challenge), "Submission limit reached for your account, it will reset at #{@reset_dttm}", :forbidden)
```
Comment (sujnesh) [inline app/controllers/submissions_controller.rb:172]: Need to add return here to avoid double render error

**Pull request merged**